File: scripts/fetch_icons.mjs
#!/usr/bin/env node
/**
 * Fetch item/equipment icon images from SchaleDB.
 *
 * Downloads master data JSON files, extracts icon references,
 * and downloads PNG icon images to assets/icons/{items,equipment}/.
 * Source images are WebP on SchaleDB; converted to PNG (with alpha) locally.
 * Idempotent: only downloads missing icons, removes stale ones.
 *
 * Usage:
 *   node scripts/fetch_icons.mjs
 *   node scripts/fetch_icons.mjs --items-only
 *   node scripts/fetch_icons.mjs --equipment-only
 *   node scripts/fetch_icons.mjs --dry-run
 */

import { writeFile, mkdir, readdir, unlink } from "node:fs/promises";
import { join, dirname } from "node:path";
import { fileURLToPath } from "node:url";
import { parseArgs } from "node:util";
import sharp from "sharp";

const __dirname = dirname(fileURLToPath(import.meta.url));
const PROJECT_ROOT = join(__dirname, "..");

const BASE_DATA_URL = "https://schaledb.com/data/jp";
const BASE_ICON_URL = "https://schaledb.com/images";
const USER_AGENT = "shittim-reader/0.1.0";
const REQUEST_INTERVAL_MS = 100;

const CATEGORIES = {
  items: {
    jsonUrl: `${BASE_DATA_URL}/items.json`,
    iconUrl: `${BASE_ICON_URL}/item/icon`,
    outputDir: "assets/icons/items",
  },
  equipment: {
    jsonUrl: `${BASE_DATA_URL}/equipment.json`,
    iconUrl: `${BASE_ICON_URL}/equipment/icon`,
    outputDir: "assets/icons/equipment",
  },
};

const sleep = (ms) => new Promise((resolve) => setTimeout(resolve, ms));
const FETCH_OPTIONS = { headers: { "User-Agent": USER_AGENT } };

async function fetchJson(url) {
  const resp = await fetch(url, FETCH_OPTIONS);
  if (!resp.ok) {
    const body = await resp.text().catch(() => "");
    throw new Error(`${url}: ${resp.status} ${resp.statusText}\n${body}`);
  }
  return resp.json();
}

async function downloadAndConvert(url, dest) {
  const resp = await fetch(url, FETCH_OPTIONS);
  if (!resp.ok) {
    console.error(`  SKIP ${url} (${resp.status})`);
    return false;
  }
  try {
    const webpBuf = Buffer.from(await resp.arrayBuffer());
    await sharp(webpBuf).png().toFile(dest);
    return true;
  } catch (err) {
    console.error(`  CONVERT FAILED ${url}: ${err.message}`);
    return false;
  }
}

async function processCategory(name, config, dryRun) {
  console.log(`\n=== ${name} ===`);

  // Fetch master data
  console.log(`Fetching ${config.jsonUrl} ...`);
  const data = await fetchJson(config.jsonUrl);

  // Save JSON for reference
  const jsonPath = join(PROJECT_ROOT, "assets", "data", `${name}.json`);
  await writeFile(jsonPath, JSON.stringify(data, null, 2), "utf-8");
  console.log(`Saved ${jsonPath} (${Object.keys(data).length} entries)`);

  // Extract unique icon names
  const icons = [
    ...new Set(
      Object.values(data)
        .map((entry) => entry.Icon)
        .filter(Boolean)
    ),
  ];
  console.log(`Found ${icons.length} unique icons`);

  if (dryRun) {
    for (const icon of icons) {
      console.log(`  [DRY RUN] ${icon}.png`);
    }
    return;
  }

  // Diff against local files
  const outputDir = join(PROJECT_ROOT, config.outputDir);
  await mkdir(outputDir, { recursive: true });

  const expectedFiles = new Set(icons.map((icon) => `${icon}.png`));
  const localFiles = new Set(await readdir(outputDir));

  const toDownload = icons.filter((icon) => !localFiles.has(`${icon}.png`));
  const toRemove = [];
  for (const f of localFiles) {
    if (f.endsWith(".webp")) toRemove.push(f);
    else if (f.endsWith(".png") && !expectedFiles.has(f)) toRemove.push(f);
  }

  // Download missing icons (WebP -> PNG conversion)
  let downloaded = 0;
  let failed = 0;

  for (const [i, icon] of toDownload.entries()) {
    if (i > 0) await sleep(REQUEST_INTERVAL_MS);
    const webpUrl = `${config.iconUrl}/${icon}.webp`;
    const pngDest = join(outputDir, `${icon}.png`);
    if (await downloadAndConvert(webpUrl, pngDest)) {
      downloaded++;
    } else {
      failed++;
    }
  }

  // Remove stale icons (parallel, no rate limiting needed)
  await Promise.all(
    toRemove.map(async (file) => {
      await unlink(join(outputDir, file));
      console.log(`  Removed stale: ${file}`);
    })
  );

  const unchanged = icons.length - toDownload.length;
  console.log(
    `Done: ${downloaded} new, ${unchanged} unchanged, ${failed} failed, ${toRemove.length} removed`
  );
}

async function main() {
  const { values } = parseArgs({
    options: {
      "items-only": { type: "boolean", default: false },
      "equipment-only": { type: "boolean", default: false },
      "dry-run": { type: "boolean", default: false },
    },
  });

  const targets = values["items-only"]
    ? ["items"]
    : values["equipment-only"]
      ? ["equipment"]
      : Object.keys(CATEGORIES);

  await mkdir(join(PROJECT_ROOT, "assets", "data"), { recursive: true });

  for (const name of targets) {
    await processCategory(name, CATEGORIES[name], values["dry-run"]);
  }

  console.log("\nAll done.");
}

main().catch((err) => {
  console.error(err);
  process.exit(1);
});

File: tools/fetch_icons.zig
//! Dev tool: download item/equipment icons from SchaleDB.
//!
//! Fetches master data JSON, extracts icon names, downloads WebP icons.
//! zigimg does not support WebP, so icons are saved as .webp files.
//!
//! Usage: zig build fetch-icons [-- --items-only | --equipment-only]

const std = @import("std");

const base_data_url = "https://schaledb.com/data/jp";
const base_icon_url = "https://schaledb.com/images";
const user_agent = "shittim-reader/0.1.0";
const request_interval_ns: u64 = 100 * std.time.ns_per_ms;

const Category = struct {
    name: []const u8,
    json_path: []const u8,
    icon_url: []const u8,
    output_dir: []const u8,
};

const categories = [_]Category{
    .{
        .name = "items",
        .json_path = "/items.json",
        .icon_url = "/item/icon",
        .output_dir = "assets/icons/items",
    },
    .{
        .name = "equipment",
        .json_path = "/equipment.json",
        .icon_url = "/equipment/icon",
        .output_dir = "assets/icons/equipment",
    },
};

pub fn main() !void {
    var gpa: std.heap.GeneralPurposeAllocator(.{}) = .init;
    defer _ = gpa.deinit();
    const allocator = gpa.allocator();

    const args = try std.process.argsAlloc(allocator);
    defer std.process.argsFree(allocator, args);

    var items_only = false;
    var equipment_only = false;
    for (args[1..]) |arg| {
        if (std.mem.eql(u8, arg, "--items-only")) items_only = true;
        if (std.mem.eql(u8, arg, "--equipment-only")) equipment_only = true;
    }

    std.fs.cwd().makePath("assets/data") catch {};

    var client: std.http.Client = .{ .allocator = allocator };
    defer client.deinit();

    for (&categories) |*cat| {
        if (items_only and !std.mem.eql(u8, cat.name, "items")) continue;
        if (equipment_only and !std.mem.eql(u8, cat.name, "equipment")) continue;
        try processCategory(allocator, &client, cat);
    }

    std.debug.print("\nAll done.\n", .{});
}

fn processCategory(allocator: std.mem.Allocator, client: *std.http.Client, cat: *const Category) !void {
    std.debug.print("\n=== {s} ===\n", .{cat.name});

    // Fetch master data JSON
    const json_url = try std.fmt.allocPrint(allocator, "{s}{s}", .{ base_data_url, cat.json_path });
    defer allocator.free(json_url);

    std.debug.print("Fetching {s} ...\n", .{json_url});

    // Download JSON to memory
    const json_body = try downloadToBuffer(allocator, client, json_url);
    defer allocator.free(json_body);

    // Save a copy to disk for reference
    const data_path = try std.fmt.allocPrint(allocator, "assets/data/{s}.json", .{cat.name});
    defer allocator.free(data_path);
    {
        const file = try std.fs.cwd().createFile(data_path, .{});
        defer file.close();
        try file.writeAll(json_body);
    }
    std.debug.print("Saved {s}\n", .{data_path});

    // Extract icon names from JSON
    var icons: std.ArrayListUnmanaged([]const u8) = .empty;
    defer {
        for (icons.items) |s| allocator.free(s);
        icons.deinit(allocator);
    }

    const parsed = try std.json.parseFromSlice(std.json.Value, allocator, json_body, .{});
    defer parsed.deinit();

    // JSON might be object or array
    const entries: []const std.json.Value = switch (parsed.value) {
        .object => |obj| blk: {
            var vals: std.ArrayListUnmanaged(std.json.Value) = .empty;
            defer vals.deinit(allocator);
            var it = obj.iterator();
            while (it.next()) |entry| {
                try vals.append(allocator, entry.value_ptr.*);
            }
            break :blk try vals.toOwnedSlice(allocator);
        },
        .array => |arr| arr.items,
        else => &.{},
    };
    defer if (parsed.value == .object) allocator.free(entries);

    for (entries) |item| {
        if (item != .object) continue;
        const icon_val = item.object.get("Icon") orelse continue;
        if (icon_val != .string) continue;
        try icons.append(allocator, try allocator.dupe(u8, icon_val.string));
    }

    std.debug.print("Found {d} icons\n", .{icons.items.len});

    std.fs.cwd().makePath(cat.output_dir) catch {};

    // Download missing icons
    var downloaded: u32 = 0;
    var skipped: u32 = 0;
    var failed: u32 = 0;

    for (icons.items) |icon_name| {
        // Save as .webp (zigimg doesn't support WebP decode)
        const webp_path = try std.fmt.allocPrint(allocator, "{s}/{s}.webp", .{ cat.output_dir, icon_name });
        defer allocator.free(webp_path);

        // Skip if already exists
        if (std.fs.cwd().access(webp_path, .{})) |_| {
            skipped += 1;
            continue;
        } else |_| {}

        if (downloaded > 0) std.Thread.sleep(request_interval_ns);

        const webp_url = try std.fmt.allocPrint(allocator, "{s}{s}/{s}.webp", .{ base_icon_url, cat.icon_url, icon_name });
        defer allocator.free(webp_url);

        downloadToFile(allocator, client, webp_url, webp_path) catch |err| {
            std.debug.print("  SKIP {s} ({s})\n", .{ icon_name, @errorName(err) });
            failed += 1;
            continue;
        };

        downloaded += 1;
    }

    std.debug.print("Done: {d} new, {d} unchanged, {d} failed\n", .{ downloaded, skipped, failed });
}

fn downloadToBuffer(allocator: std.mem.Allocator, client: *std.http.Client, url: []const u8) ![]u8 {
    const uri = try std.Uri.parse(url);
    var req = try client.request(.GET, uri, .{
        .headers = .{
            .accept_encoding = .{ .override = "identity" },
        },
        .extra_headers = &.{
            .{ .name = "User-Agent", .value = user_agent },
        },
    });
    defer req.deinit();
    try req.sendBodiless();

    var redirect_buf: [8192]u8 = undefined;
    var response = try req.receiveHead(&redirect_buf);

    if (response.head.status != .ok) {
        return error.HttpError;
    }

    var transfer_buf: [16384]u8 = undefined;
    const body_reader = response.reader(&transfer_buf);

    var result: std.ArrayListUnmanaged(u8) = .empty;
    errdefer result.deinit(allocator);

    var chunk: [16384]u8 = undefined;
    while (true) {
        const n = try body_reader.read(&chunk);
        if (n == 0) break;
        try result.appendSlice(allocator, chunk[0..n]);
    }
    return try result.toOwnedSlice(allocator);
}

fn downloadToFile(allocator: std.mem.Allocator, client: *std.http.Client, url: []const u8, dest_path: []const u8) !void {
    const data = try downloadToBuffer(allocator, client, url);
    defer allocator.free(data);

    const file = try std.fs.cwd().createFile(dest_path, .{});
    defer file.close();
    try file.writeAll(data);
}
